### PR TITLE
lib/model, lib/protocol: Wait for reader/writer loops on close (fixes #4170)

### LINKED ISSUE #4170: panic: bug: ClusterConfig called on closed or nonexistent connection
I have had two different Syncthing instances crash on Windows 10. One is running v0.14.28, one v.0.14.29-rc1.
Both are using SyncTrayzor (V1.1.16).

I restarted one, only to have it crash ~5 mins later. I captured the Synctrazor log from that one and have attached it.

I will endeavour to provide any other information required.

Forum post for reference: [https://forum.syncthing.net/t/syncthing-crashing/9845](url)

[SyncTrayzor.txt](https://github.com/syncthing/syncthing/files/1021902/SyncTrayzor.txt)

### PR DESCRIPTION
**Pull request by imsodin**
### Purpose

Fix the infamous cluster config panics #4170.

### Testing

None, as the panic isn't reliably reproducible, but it am confident xD

### CODE REVIEW

**Comment by AudriusButkevicius** [comment]
Wait group might be more simple?

**Comment by AudriusButkevicius** [comment]
I wonder if there is no deadlocks here.

Reader blocks on delivering a message, which tries to acquire a pmut/fmut in the model, and the conn.Close() happens under pmut/fmut which is blocked on the wg.wait returning, which is blocked on reader loop exiting.

**Comment by imsodin** [comment]
That's indeed a possibility. Solution: Don't hold the locks in model while closing.

And because I couldn't help myself I cleaned up a little bit of legacy code in the touched areas.

File: lib/model/model.go [hunk 1]
```diff
@@ -1412,21 +1407,21 @@ func (m *model) Closed(conn protocol.Connection, err error) {
...
 		// There is no connection to close
 		return
 	}
-
 	conn.Close(err)
 }
 
+func (m *model) closeConnToDevices(fcfg config.FolderConfiguration, err error) {
+	for _, dev := range fcfg.DeviceIDs() {
+		m.close(dev, err)
```
Comment (AudriusButkevicius) [inline lib/model/model.go]: This acquires pmut repeatedly.

File: lib/model/model.go [hunk 2]
```diff
@@ -1412,21 +1407,21 @@ func (m *model) Closed(conn protocol.Connection, err error) {
...
+	m.pmut.Unlock()
 	if !ok {
 		// There is no connection to close
 		return
 	}
-
 	conn.Close(err)
 }
 
+func (m *model) closeConnToDevices(fcfg config.FolderConfiguration, err error) {
```
Comment (AudriusButkevicius) [inline lib/model/model.go]: This could just take []DeviceID as that is what DeviceIDs() returns.

**Pull request merged**